staging/src/k8s.io/apiserver/pkg/endpoints/request/requestcontext_test.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"net/http"
+	"sync"
+	"testing"
+)
+
+func TestRequestContextMapperGet(t *testing.T) {
+	mapper := NewRequestContextMapper()
+	context := NewContext()
+	req, _ := http.NewRequest("GET", "/api/version/resource", nil)
+
+	// empty mapper
+	if _, ok := mapper.Get(req); ok {
+		t.Fatalf("got unexpected context")
+	}
+
+	// init mapper
+	mapper.(*requestContextMap).init(req, context)
+	if _, ok := mapper.Get(req); !ok {
+		t.Fatalf("got no context")
+	}
+
+	// remove request context
+	mapper.(*requestContextMap).remove(req)
+	if _, ok := mapper.Get(req); ok {
+		t.Fatalf("got unexpected context")
+	}
+
+}
+func TestRequestContextMapperUpdate(t *testing.T) {
+	mapper := NewRequestContextMapper()
+	context := NewContext()
+	req, _ := http.NewRequest("GET", "/api/version/resource", nil)
+
+	// empty mapper
+	if err := mapper.Update(req, context); err == nil {
+		t.Fatalf("got no error")
+	}
+
+	// init mapper
+	if !mapper.(*requestContextMap).init(req, context) {
+		t.Fatalf("unexpected error, should init mapper")
+	}
+
+	context = WithNamespace(context, "default")
+	if err := mapper.Update(req, context); err != nil {
+		t.Fatalf("unexpected error")
+	}
+
+	if context, ok := mapper.Get(req); !ok {
+		t.Fatalf("go no context")
+	} else {
+		if ns, _ := NamespaceFrom(context); ns != "default" {
+			t.Fatalf("unexpected namespace %s", ns)
+		}
+	}
+}
+
+func TestRequestContextMapperConcurrent(t *testing.T) {
+	mapper := NewRequestContextMapper()
+
+	testCases := []struct{ url, namespace string }{
+		{"/api/version/resource1", "ns1"},
+		{"/api/version/resource2", "ns2"},
+		{"/api/version/resource3", "ns3"},
+		{"/api/version/resource4", "ns4"},
+		{"/api/version/resource5", "ns5"},
+	}
+
+	wg := sync.WaitGroup{}
+	for _, testcase := range testCases {
+		wg.Add(1)
+		go func(testcase struct{ url, namespace string }) {
+			defer wg.Done()
+			context := NewContext()
+			req, _ := http.NewRequest("GET", testcase.url, nil)
+
+			if !mapper.(*requestContextMap).init(req, context) {
+				t.Fatalf("unexpected init error")
```

Comment:
`t.Fatalf()` calls `t.FailNow()` which has the following comment:
```
// FailNow marks the function as having failed and stops its execution
// by calling runtime.Goexit (which then runs all deferred calls in the
// current goroutine).
// Execution will continue at the next test or benchmark.
// FailNow must be called from the goroutine running the
// test or benchmark function, not from other goroutines
// created during the test. Calling FailNow does not stop
// those other goroutines.
func (c *common) FailNow() {
```
So it is not right to call it here. I think this (and below) should be `Errorf()` and then `return`.

Reply:
I do not see much difference.

Reply:
I have created https://github.com/golang/go/issues/24678 to see what Go devs think about catching this issue automatically (when a test fails...).